wiki/C++-Style-Guide.md

```diff
@@ -0,0 +1,814 @@
+
+- Readability is achieved by laying out text in a clear visual hierarchy.
+- Code should not rely on features of a particular editor or compiler.
+- Formatting should be consistent to avoid confusing or distracting the reader.
+- Reducing the risk of bugs is important even if it takes more work.
+
+
+# Files
+
+This section deals with naming and encoding of source code files and the items that appear at the start of each file.
+
+## Line endings
+
+All files should have Unix-style line endings.
+That is, "\n" alone, not "\r\n" (Windows style). If an editor changes Unix line endings to Windows style, this causes issues with source control diffs because those lines appear to be changed even though their actual text is the same.
+
+## File encodings
+
+All files should be plain ASCII.
+That is, nothing beyond the 127-character ASCII set should be used. This is to ensure maximum cross-compatibility.
+
+## File extensions
+
+All C++ files should use ".h" and ".cpp" for their extensions.
+This is a somewhat arbitrary choice, but these seem to be the most commonly used extensions.
+
+## One class per file
+
+Each class "MyClass" should have its own .h and .cpp files, and they should be named "MyClass.h" and "MyClass.cpp".
+This is so that it is easy to find the implementation of any class you are interested in.
+
+## Names for files with no classes
+
+If a file does not contain any classes (such as "main.cpp" or "shift.h"), its name should be lower-case.
+This is just to make it clear that the file is not defining a class.
+
+## Header file guards
+
+All header files should have #define guards, and they should be in the format "#ifndef MY_CLASS_H_".
+The C++ standard technically does not allow source code files to define macros that start with an underscore or contain a double underscore.
+
+## Header and copyright statement
+
+Each file should begin with a header which specifies the file name, copyright, and licensing information, e.g.:
+```c++
+/* Key.h
+Copyright (c) 2014 by Michael Zahniser
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+```
+If you make substantial additions or changes to a file, please add your name to the copyright list.
```

Comment:
Can you elaborate on what it would mean to do them properly?
As I understand it, the current approach is indeed for people who make substantial changes to a file, or create new code in a new file, to add their name to the copyright at the beginning, though this is perhaps more common in relation to data files than it is to the source code.

Reply:
As far as I'm aware, people's names only get put in the copyright header when they're the one who makes it, for ES. At least, that's the rules I've been playing by. I think, according to the GPL license, _every single_ contributor to the file should have their name in the header. We definitely aren't doing that.